c_src/ex_dtls/native.c

```diff
@@ -92,90 +128,136 @@ UNIFEX_TERM do_handshake(UnifexEnv *env, State *state, UnifexPayload *payload) {
     DEBUG("RBIO: wrote %d", bytes);
   }
 
-  for (;;) {
-    int res = SSL_do_handshake(state->ssl);
-
-    size_t pending_data_len = BIO_ctrl_pending(SSL_get_wbio(state->ssl));
-    if (pending_data_len > 0) {
-      DEBUG("WBIO: pending data: %ld bytes", pending_data_len);
+  char data[1500] = {0};
+  int ret = SSL_read(state->ssl, &data, 1500);
+
+  if (state->hsk_finished == 1) {
+    return handle_regular_read(state, data, ret);
+  } else if (SSL_is_init_finished(state->ssl) == 1) {
+    DEBUG("Handshake finished successfully");
+    return handle_handshake_finished(state);
+  } else {
+    DEBUG("Handshake in progress");
+    return handle_handshake_inprogress(state, ret);
+  }
+}
 
-      char *data = (char *)malloc(pending_data_len * sizeof(char));
-      memset(data, 0, pending_data_len);
-      BIO *wbio = SSL_get_wbio(state->ssl);
-      int read_bytes = BIO_read(wbio, data, pending_data_len);
-      if (read_bytes <= 0) {
-        DEBUG("WBIO: read error");
-        return unifex_raise(state->env, "Handshake failed: write BIO error");
-      } else {
-        if (dyn_buff_insert(dyn_buff, data, read_bytes) == -1) {
-           return unifex_raise(state->env, "Handshake failed: can't insert to dyn_buff");
-        }
-      }
+UNIFEX_TERM handle_regular_read(State *state, char data[], int ret) {
+  if (ret > 0) {
+    UnifexPayload *packets = (UnifexPayload *)unifex_payload_alloc(
+        state->env, UNIFEX_PAYLOAD_BINARY, ret);
+    memcpy(packets->data, data, ret);
+    packets->size = (unsigned int)ret;
+    UNIFEX_TERM res_term = process_result_ok(state->env, state, packets);
+    unifex_payload_release(packets);
+    return res_term;
+  }
 
-      DEBUG("WBIO: read: %d bytes", read_bytes);
+  // handle READ errors including DTLS alerts
+  int error = SSL_get_error(state->ssl, ret);
+  DEBUG("SSL ERROR: %d", error);
+  return process_result_error(state->env, error);
```

Comment:
I am not sure how to handle it. We can:
* raise here
* return error with error value but error value will mean nothing for the user
* return :ok and if DTLS alert occurred we can send message from `ssl_info_cb`

`error` can mean DTLS alerts but also some other retryable or non retryable errors. I would leave it as it is and see how it behaves.

Reply:
In general, we should raise on critical, non-retryable errors, warn on non-critical errors and retry on the retryable ones ;) If we don't know which errors are critical, let's always warn, at least for now.